xarray/core/duck_array_ops.py

```diff
@@ -217,23 +217,27 @@ def asarray(data, xp=np):
     return data if is_duck_array(data) else xp.asarray(data)
 
 
-def as_shared_dtype(scalars_or_arrays, xp=np):
-    """Cast a arrays to a shared dtype using xarray's type promotion rules."""
-    array_type_cupy = array_type("cupy")
-    if array_type_cupy and any(
-        isinstance(x, array_type_cupy) for x in scalars_or_arrays
-    ):
-        import cupy as cp
-
-        arrays = [asarray(x, xp=cp) for x in scalars_or_arrays]
+def as_duck_array(data, xp=np):
+    if is_duck_array(data):
+        return data
+    elif hasattr(data, "get_duck_array"):
+        # must be a lazy indexing class wrapping a duck array
+        return data.get_duck_array()
```

Comment:
Will this idea always work? What if it steps down through a lazy decoder class that changes the dtype...

Reply:
Those should be going through https://github.com/pydata/xarray/blob/c9ba2be2690564594a89eb93fb5d5c4ae7a9253c/xarray/coding/variables.py#L52-L64

so you should be fine.

Reply:
I think I'm getting confused as to how this all works now... Don't I want to be computing `as_shared_dtype` using the dtype of the outermost wrapped class? Whereas this will step through all the way to the innermost duckarray, which may have a different dtype?

Reply:
As of now, `as_shared_dtype` is expected to return pure duck arrays: stack, concatenate, and where.

So that means we need to read from disk, which you do with `to_duck_array` and all these wrapper layers will be resolved.